Mathlib/Data/LazyList/Basic.lean

```diff
@@ -86,98 +69,32 @@ instance : LawfulTraversable LazyList := by
-  | LazyList.nil, _ => LazyList.nil
-  | LazyList.cons x xs, f => (f x).append (xs.get.bind f)
 #align lazy_list.bind LazyList.bind
-
-/-- Reverse the order of a `LazyList`.
-It is done by converting to a `List` first because reversal involves evaluating all
-the list and if the list is all evaluated, `List` is a better representation for
-it than a series of thunks. -/
-def reverse {α} (xs : LazyList α) : LazyList α :=
-  ofList xs.toList.reverse
 #align lazy_list.reverse LazyList.reverse
-
-instance : Monad LazyList where
-  pure := @LazyList.singleton
-  bind := @LazyList.bind
-
--- Porting note: Added `Thunk.pure` to definition.
-theorem append_nil {α} (xs : LazyList α) : xs.append (Thunk.pure LazyList.nil) = xs := by
-  induction' xs using LazyList.rec with _ _ _ _ ih
-  · simp only [Thunk.pure, append, Thunk.get]
-  · simpa only [append, cons.injEq, true_and]
-  · ext; apply ih
 #align lazy_list.append_nil LazyList.append_nil
-
-theorem append_assoc {α} (xs ys zs : LazyList α) :
-    (xs.append ys).append zs = xs.append (ys.append zs) := by
-  induction' xs using LazyList.rec with _ _ _ _ ih
-  · simp only [append, Thunk.get]
-  · simpa only [append, cons.injEq, true_and]
-  · ext; apply ih
 #align lazy_list.append_assoc LazyList.append_assoc
-
--- Porting note: Rewrote proof of `append_bind`.
-theorem append_bind {α β} (xs : LazyList α) (ys : Thunk (LazyList α)) (f : α → LazyList β) :
-    (xs.append ys).bind f = (xs.bind f).append (ys.get.bind f) := by
-  match xs with
-  | LazyList.nil =>
-    simp only [append, Thunk.get, LazyList.bind]
-  | LazyList.cons x xs =>
-    simp only [append, Thunk.get, LazyList.bind]
-    have := append_bind xs.get ys f
-    simp only [Thunk.get] at this
-    rw [this, append_assoc]
 #align lazy_list.append_bind LazyList.append_bind
 
+@[simp] theorem bind_singleton {α} (x : LazyList α) : x.bind singleton = x := by
+  induction x using LazyList.rec (motive_2 := fun xs => xs.get.bind singleton = xs.get) with
+  | nil => simp [LazyList.bind]
+  | cons h t ih =>
+    simp only [LazyList.bind, singleton, append, Thunk.get_pure, Thunk.get_mk, cons.injEq, true_and]
+    ext
+    simp [ih]
+  | mk f ih => simp_all
+
 instance : LawfulMonad LazyList := LawfulMonad.mk'
-  (bind_pure_comp := by
-    intro _ _ f xs
-    simp only [bind, Functor.map, pure, singleton]
-    induction xs using LazyList.traverse.induct (m := @Id) (f := f) with
-    | case1 =>
-      simp only [Thunk.pure, LazyList.bind, LazyList.traverse, Id.pure_eq]
-    | case2 _ _ ih =>
-      simp only [LazyList.bind, LazyList.traverse, Seq.seq, Id.map_eq, append, Thunk.pure]
-      rw [← ih]
-      simp [Thunk.pure, Thunk.get, append])
+  (id_map := by
+    intro α xs
+    induction xs using LazyList.rec (motive_2 := fun xs => id <$> xs.get = xs) with
+    | nil => simp only [Functor.map, comp_id, LazyList.bind]
+    | cons h t _ => simp only [Functor.map, comp_id, bind_singleton]
+    | mk f _ => ext; simp_all)
```

Comment:
Any particular reason you swapped the order of these?

Reply:
Matches the order of the arguments.